physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,718 @@
+		CleanupDeadServers:             false,
+		LastContactThreshold:           10 * time.Second,
+		DeadServerLastContactThreshold: 24 * time.Hour,
+		MaxTrailingLogs:                1000,
+		MinQuorum:                      3,
+		ServerStabilizationTime:        10 * time.Second,
+	}
+}
+
+func (b *RaftBackend) AutopilotDisabled() bool {
+	b.l.RLock()
+	disabled := b.disableAutopilot
+	b.l.RUnlock()
+	return disabled
+}
+
+// AutopilotExecutionStatus represents the current status of the autopilot background go routines
+type AutopilotExecutionStatus string
+
+const (
+	AutopilotNotRunning   AutopilotExecutionStatus = "not-running"
+	AutopilotRunning      AutopilotExecutionStatus = "running"
+	AutopilotShuttingDown AutopilotExecutionStatus = "shutting-down"
+)
+
+func autopilotStatusToStatus(status autopilot.ExecutionStatus) AutopilotExecutionStatus {
+	switch status {
+	case autopilot.Running:
+		return AutopilotRunning
+	case autopilot.ShuttingDown:
+		return AutopilotShuttingDown
+	default:
+		return AutopilotNotRunning
+	}
+}
+
+func (b *RaftBackend) startFollowerHeartbeatTracker() {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-b.followerHeartbeatTrackerStopCh:
+			return
+		case <-ticker.C:
+			markingCandidatePresent := false
+			b.l.RLock()
+			if b.autopilotConfig.CleanupDeadServers && b.autopilotConfig.DeadServerLastContactThreshold != 0 {
+				b.followerStates.l.RLock()
+				for _, state := range b.followerStates.followers {
+					if state.LastHeartbeat.IsZero() || state.IsDead {
+						continue
+					}
+					now := time.Now()
+					if now.After(state.LastHeartbeat.Add(b.autopilotConfig.DeadServerLastContactThreshold)) {
+						markingCandidatePresent = true
+					}
+				}
+				b.followerStates.l.RUnlock()
+
+				// If marking a node as dead, switch read with a write lock and
+				// check the conditions again.
+				if markingCandidatePresent {
+					b.followerStates.l.Lock()
+					for id, state := range b.followerStates.followers {
+						if state.LastHeartbeat.IsZero() || state.IsDead {
+							continue
+						}
+						now := time.Now()
+						threshold := state.LastHeartbeat.Add(b.autopilotConfig.DeadServerLastContactThreshold)
+						if now.After(threshold) {
+							b.logger.Info("marking node as dead", "last_heartbeat", state.LastHeartbeat.String(), "dead_server_last_contact_threshold", b.autopilotConfig.DeadServerLastContactThreshold, "now", now, "threshold", threshold)
+							b.followerStates.markFollowerAsDead(id)
+
+						}
+					}
+					b.followerStates.l.Unlock()
+				}
+			}
+			b.l.RUnlock()
+		}
+	}
+}
+
+// StopAutopilot stops a running autopilot instance. This should only be called
+// on the active node.
+func (b *RaftBackend) StopAutopilot() {
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	if b.autopilot == nil {
+		return
+	}
+	b.autopilot.Stop()
+
+	if b.followerHeartbeatTrackerStopCh != nil {
+		close(b.followerHeartbeatTrackerStopCh)
+	}
+	b.followerHeartbeatTrackerStopCh = nil
```

Comment:
Is it wise to set this nil?  For one thing, it seems racy.  For another, reads from nil channels block forever, so it could prevent the startFollowerHeartbeatTracker goroutine from exiting.

Reply:
This should be okay. We take this approach with `raftTLSRotationStopCh` as well. The main reason to do this is that, for reasons unknown to me yet, `StopAutopilot` gets called more than once which leads to a "closed channel getting closed agiain"-ish error. I figured the `raftTLSRotationStopCh` also had the same issue and it was circumvented by this approach. IMO, we shouldn't land in a scenario where we are reading from nil channel.

Reply:
I believe you're seeing StopAutopilot called multiple times because preSeal gets called when we stop being the leader, as well as when we're about to seal.

I'm not convinced that this is ok: right after we close we set to nil, so the select in startFollowerHeartbeatTracker is unlikely to ever return. 

As an alternative, you could consider replacing the backend field followerHeartbeatTrackerStopCh with the ticker that the goroutine ranges over, and calling Stop on the ticker in StopAutopilot.  I think it's safe to call that Stop twice: https://play.golang.org/p/LWF1C7o-_kP

Reply:
Re the racy assignment: https://app.circleci.com/pipelines/github/hashicorp/vault/15693/workflows/88f6b46e-c356-4605-9823-04821b0f60e1/jobs/160346

Reply:
Loved the suggestion. Done.